libraries/Microsoft.Bot.Builder.Core/Middleware/StateMiddleware.cs

```diff
@@ -0,0 +1,174 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.Bot.Builder.Middleware;
+using Microsoft.Bot.Schema;
+
+namespace Microsoft.Bot.Builder.Middleware
+{
+    public class StateSettings
+    {
+        public bool WriteBeforeSend { get; set; } = true;
+        public bool LastWriterWins { get; set; } = true;
+    }
+
+    public abstract class StateMiddleware<StateT> : IContextCreated, ISendActivity
+        where StateT : IStoreItem, new()
+    {
+        private readonly StateSettings _settings;
+        private readonly IStorage _storage;
+
+        /// <summary>
+        /// Create statemiddleware
+        /// </summary>
+        /// <param name="name">name of the kind of state</param>
+        /// <param name="storage">storage provider to use</param>
+        /// <param name="settings">settings</param>
+        public StateMiddleware(IStorage storage, StateSettings settings = null)
+        {
+            _storage = storage ?? throw new ArgumentNullException(nameof(storage));
+            _settings = settings ?? new StateSettings();
+        }
+
+        public abstract string GetPropertyName();
+
+        /// <summary>
+        /// The key extracted from the activity which is used to store/retrieve the state
+        /// </summary>
+        /// <param name="activity"></param>
+        /// <returns></returns>
+        public abstract string GetStorageKey(IBotContext context);
+
+        public async Task ContextCreated(IBotContext context, MiddlewareSet.NextDelegate next)
+        {
+            await Read(context).ConfigureAwait(false);
+            await next().ConfigureAwait(false);
+        }
+
+        public async Task SendActivity(IBotContext context, IList<Activity> activities, MiddlewareSet.NextDelegate next)
+        {
+            if (_settings.WriteBeforeSend)
+            {
+                await Write(context).ConfigureAwait(false);
+            }
+            await next().ConfigureAwait(false);
+            if (!_settings.WriteBeforeSend)
+            {
+                await Write(context).ConfigureAwait(false);
+            }
+        }
+
+        protected virtual async Task<StoreItems> Read(IBotContext context)
```

Comment:
>Read(IBotContext context) [](start = 49, length = 25)

my signature for this is read(context, force = false) and my implementation is a little different.

My version of read will first check to see if the state has been cached on the context and only read it in if it's not there. That way you can call read() as many times as you want without penalty. If you pass "force = true" then it will force a re-read of the cached value.  I have a similar flag on write() which by default will only write out the state if it's changed. Passing "force = true" to write causes it to write anyway.

These are subtle differences but it means you can create a ConversationState instance and use it free-standing outside of middleware. You would just call read() and write() directly from within your code.  You're's supports that too but makes the bots logic check to see if the state has been loaded and cached before calling read().